fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,112 @@
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    *   - create a shallow copy of the todos array from state
+    *   - add the todo received from the server to the copied array
+    *   - replace the old todos array in state with the new one
+    *
+    * DO NOT MUTATE the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newTodo)
+    })
+      .then(res => res.json())
+      .then(data => this.setState({
+        todos: this.state.todos.concat(data)
+      }))
+      .catch(err => console.error(err));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing ONE PROPERTY: the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     *   - create a shallow copy of the todos array from state
+     *   - replace the old todo with the todo received from the server
+     *   - replace the old todos in the state with the new one (you know the index).
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list repeatedly should "toggle" its isCompleted status back and forth.
+     *
+     * DO NOT try to calculate the index of the todo by subtracting 1 from the id.
+     *
+     * DO NOT MUTATE the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    const todo = {};
+    for (let i = 0; i < this.state.todos.length; i++) {
+      if (this.state.todos[i].todoId === todoId) {
+        todo.isCompleted = !this.state.todos[i].isCompleted;
+        fetch(`/api/todos/${todoId}`, {
+          method: 'PATCH',
+          body: JSON.stringify(todo),
+          headers: { 'Content-Type': 'application/json' }
+        })
+          .then(res => res.json())
+          .then(data => {
+            const todosCopy = [...this.state.todos];
+            todosCopy[i] = data;
+            this.setState({ todos: todosCopy });
+          })
+          .catch(err => console.error(err));
+      }
+    }
```

Comment:
While this looks like it works it is best to not make the `fetch` request within a `for` loop. It is better to first find the `index` then make the request after the loop. Also instead of using a `for` loop you can use something like [findIndex](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex).

For example:
```JAVASCRIPT
const index = this.state.todos.findIndex(todo => todo.todoId === todoId);
const todo = { isCompleted: !this.state.todos[index].isCompleted };

fetch(`/api/todos/${todoId}`, {
  method: 'PATCH',
  body: JSON.stringify(todo),
  headers: { 'Content-Type': 'application/json' }
})
  .then(res => res.json())
  .then(data => {
    const todosCopy = [ ...this.state.todos ];
    todosCopy[index] = data;
    this.setState({ todos: todosCopy });
  })
  .catch(err => console.error(err));
```